src/common/fiber.cpp

```diff
@@ -116,16 +116,21 @@ void Fiber::Rewind() {
     boost::context::detail::jump_fcontext(impl->rewind_context, this);
 }
 
-void Fiber::YieldTo(std::shared_ptr<Fiber> from, std::shared_ptr<Fiber> to) {
-    ASSERT_MSG(from != nullptr, "Yielding fiber is null!");
+void Fiber::YieldTo(std::weak_ptr<Fiber> weak_from, std::shared_ptr<Fiber> to) {
     ASSERT_MSG(to != nullptr, "Next fiber is null!");
+
     to->impl->guard.lock();
-    to->impl->previous_fiber = from;
+    to->impl->previous_fiber = weak_from.lock();
+
     auto transfer = boost::context::detail::jump_fcontext(to->impl->context, to.get());
-    ASSERT(from->impl->previous_fiber != nullptr);
-    from->impl->previous_fiber->impl->context = transfer.fctx;
-    from->impl->previous_fiber->impl->guard.unlock();
-    from->impl->previous_fiber.reset();
+
+    // "from" might no longer be valid if the thread was killed
+    if (auto from = weak_from.lock(); from != nullptr) {
```

Comment:
This can be simplified to:

```suggestion
    if (auto from = weak_from.lock()) {
```